miden-proto/proto/rpc.proto

```diff
@@ -0,0 +1,74 @@
+// Specification of the user facing gRPC API.
+//
+// This server is a thin layer over the other services, it will mostly forward
+// requests downstream.
+syntax = "proto3";
+package rpc;
+
+import "account.proto";
+import "block.proto";
+import "digest.proto";
+import "note.proto";
+import "transaction.proto";
+
+message GetBlockHeaderRequest {
+    block.BlockSpec block_spec = 1;
+}
+
+message GetNoteRequest {
+    note.NoteSpec note_spec = 2;
+}
+
+message GetTransactionDataRequest {
+    account.AccountId account_id = 1;
+    block.BlockSpec block_spec = 2;
+    repeated note.NoteOrigin notes = 3;
+}
+
+message ListNotesByTagRequest {
+    // Only include tags after this block.
+    block.BlockSpec block_spec = 1;
+    // Minimum value of the tag.
+    fixed64 tag_min = 2;
+    // Maximum value of the tag.
+    fixed64 tag_max = 3;
+}
+
+message SubmitTransactionRequest {
+    transaction.ProvenTransaction transaction = 1;
+}
+
+// Server response. The format of the data depends on the endpoint, it is the
+// responsability of the client to decode the data form its original format.
+message ResponseBytes {
+    bytes data = 1;
+}
+
+service Api {
+    // QUERY ENPOINTS
+    // -------------------------------------------------------------------------------------------
+
+    // Query a block header by block spec.
+    rpc GetBlockHeader(GetBlockHeaderRequest) returns (ResponseBytes) {}
+
+    // Query a note by note spec.
+    rpc GetNote(GetNoteRequest) returns (ResponseBytes) {}
+
+    // Query for nullifiers.
+    rpc CheckNullifiers(stream digest.Digest) returns (stream ResponseBytes) {}
```

Comment:
A general question: is there a way to control the size of the incoming stream? For example, let's say we want to put an upper limit on how many digests can be submitted with this request - is there a way to do it?

Another general thought: we may want to make the responses a bit more structured sometimes. Like in this case we know it will be a list of nullifiers and their Merkle paths. So, the response could be something like `stream NullifierInfo`, where `NullifierInfo` could be something like:
```proto
message NulllifierInfo {
    Digest nullifier = 1;
    bytes merkle_path = 2;
}
```
Though, I guess this will depend on how we store this data. For example, in the above, I'm assuming that nullifier values and their Merkle paths exist separately.

Reply:
> A general question: is there a way to control the size of the incoming stream?

I don't think we can control the size of the incoming stream. But it should be doable to control how much data the server accepts, and to control when to close the connection.

> For example, let's say we want to put an upper limit on how many digests can be submitted with this request - is there a way to do it?

This should be doable. If this is the goal, I recon we will need pagination for this endpoint.

> Another general thought: we may want to make the responses a bit more structured sometimes. Like in this case we know it will be a list of nullifiers and their Merkle paths

The responses were all made bytes because of the previous request. We can change it back if necessary.

Reply:
> I don't think we can control the size of the incoming stream. But it should be doable to control how much data the server accepts, and to control when to close the connection.

Yeah - I saw that we could set a maximum message size - though I wonder if it affects both requests and responses and also how it works in the context of streaming. Ideally, we'd be able to limit request size by endpoint - but this is not something we need to do now. Just something to keep in mind.

> The responses were all made bytes because of the previous request. We can change it back if necessary.

Yeah - I think as we work through endpoints we can settle on specific shapes. In some cases, it may make sense to add a bit more structure to them.

Reply:
> though I wonder if it affects both requests and responses and also how it works in the context of streaming

> Currently, both servers and clients can be configured to set the max message encoding and decoding size. This will ensure that an incoming gRPC message will not exahust the systems memory. By default, the decoding message limit is 4MB and the encoding limit is usize::MAX.

https://docs.rs/tonic/latest/tonic/#max-message-size

> Ideally, we'd be able to limit request size by endpoint - but this is not something we need to do now. Just something to keep in mind.

It should be doable, we would just need to make each endpoint a unique rpc interface